### DISCUSSION TITLE
Use TryAdd for registering services?

### DISCUSSION THREAD

**Discussion by strohhut** [Dotnet-Boxed/Templates#135] [Q&A]
I think this might be better in case a service has already been registered. Unfortunately there are no generic overloads.

**Comment by RehanSaeed** [comment]
What is already registered that you are trying to register?

**Comment by strohhut** [comment]
I don't have a specific case, but when somebody uses your Startup.cs from the template and also adds some services on his own it might be better if at least the template stuff uses  `TryAdd` in order to lower the chance of registering a service multiple times.

**Comment by RehanSaeed** [comment]
I assume that doing this would cause an exception and then they could remove the line that they added because it's not needed.

**Comment by strohhut** [comment]
I don't think exceptions are thrown. I tried

``` CSharp
services.AddSingleton<IClass1, Class1>(provider => new Class1());
services.AddSingleton<IClass1, Class1>(provider => new Class1());
services.AddSingleton<IClass1, Class1B>(provider => new Class1B());
```

and only the constructor from `Class1B` was run.

Now when I have 

``` CSharp
services.AddSingleton<IClass1, Class1>(provider => new Class1());
services.AddSingleton<IClass1, Class1>(provider => new Class1());
services.TryAddSingleton(typeof(IClass1), provider => new Class1B());
```

only the `Class1` constructor is executed once.

**Comment by RehanSaeed** [comment]
Right, I think by default if the service is already there it gets replaced, the TryAdd only adds it if it's not already been added. You are worried that someone will override one of the services that have been registered and it will get replaced. I don't think there would be any bad consequences.

**Comment by strohhut** [comment]
Yeah, that's what I'm thinking. 

Maybe a custom `IUrlHelper` that always appends some query string parameters for generated URLs if a user is authenticated.

I don't have any good use cases. But maybe using `TryAdd` for the Startup.cs registrations would make the template a bit more defensive.

**Comment by RehanSaeed** [comment]
Unfortunately, we can only do so much to stop developers shooting themselves in the foot. The backlog on this project is getting a little large, without looking at an edge case like this.
